src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TabKey.cs

```diff
@@ -67,60 +64,91 @@ private bool TryInvokeSnippetCompletion(TabKeyCommandArgs args, CompletionServic
-                            this.StartNewModelComputation(completionService, new CompletionTrigger(CompletionTriggerKind.Snippets), filterItems: false);
-                            return true;
-                        }
-                    }
-                }
+                return false;
+            }
+
+            var documentId = workspace.GetDocumentIdInCurrentContext(subjectBuffer.AsTextContainer());
+            if (documentId == null)
+            {
+                return false;
+            }
+
+            var document = workspace.CurrentSolution.GetDocument(documentId);
+            if (document == null)
+            {
+                return false;
+            }
+
+            var rules = GetCompletionService().GetRules();
+            if (rules.SnippetsRule != SnippetsRule.IncludeAfterTypingIdentifierQuestionTab)
+            {
+                return false;
             }
 
-            return false;
+            var syntaxFactsOpt = document.GetLanguageService<ISyntaxFactsService>();
+            if (syntaxFactsOpt == null ||
+                caretPoint < 2 ||
+                text[caretPoint - 1] != '?' ||
+                !QuestionMarkIsPrecededByIdentifierAndWhitespace(text, caretPoint - 1, syntaxFactsOpt))
+            {
+                return false;
+            }
+
+            // Because <question><tab> is actually a command to bring up snippets,
+            // we delete the last <question> that was typed.
+            var textChange = new TextChange(TextSpan.FromBounds(caretPoint - 1, caretPoint), string.Empty);
+            workspace.ApplyTextChanges(documentId, textChange, CancellationToken.None);
+            this.StartNewModelComputation(completionService, new CompletionTrigger(CompletionTriggerKind.Snippets), filterItems: false);
+            return true;
         }
 
-        private bool QuestionMarkIsPrecededByIdentifierAndWhitespace(SourceText text, int p, ISyntaxFactsService syntaxFacts)
+        private bool QuestionMarkIsPrecededByIdentifierAndWhitespace(
+            SourceText text, int questionPosition, ISyntaxFactsService syntaxFacts)
         {
-            int start = text.Lines.GetLineFromPosition(p).Start;
-            bool seenIdentifier = false;
+            var startOfLine = text.Lines.GetLineFromPosition(questionPosition).Start;
 
-            while (p >= start)
+            // First, skip all the whitespace.
+            var current = startOfLine;
+            while (current < questionPosition && char.IsWhiteSpace(text[current]))
             {
-                if (!(syntaxFacts.IsIdentifierStartCharacter(text[p]) || syntaxFacts.IsIdentifierPartCharacter(text[p])))
-                {
-                    break;
-                }
-
-                seenIdentifier = true;
-                p--;
+                current++;
             }
 
-            while (p >= start)
+            if (current < questionPosition && syntaxFacts.IsIdentifierStartCharacter(text[current]))
```

Comment:
Is this test trying to avoid running the loop below in the case where the first non-whitespace character on the line is not an identifier char?